### ISSUE TITLE
Multi-GPU support?

### ISSUE THREAD

**Issue by harrydet** [aparapi/aparapi#7]
Hello, 

There is a thread on the Google code website talking about multi-GPU support. A comment made by a contributor to this project suggest that the Range class contains a method called setDevice(Device d) which allows the range to be executed on a specific GPU or device in general. I haven't been able to find this method however. Does it exist in the available versions? If so, how can I find it.

Thank you

**Comment by grfrost** [comment]
By default you will get whatever Aparapi conisders to be the 'best' OpenCL device.  

However  if you create your 'Range' from a device then you can select the device. 

Even though this page says it was a proposal.  I do think this was implemented 
https://code.google.com/p/aparapi/wiki/DeviceProposal

So .. something like...  

 Device chosen=null;
  for (Device device: devices.getAll()){
     if (device.getVendor().contains("AMD") && device.isGPU()){
        chosen = device;
        break;
     }
  } 
range = device.createRange2D(1024, 1024);
kernel.execute(range);

But I answered this without referencing code, so API's may be slightly different. 

Gary

**Comment by harrydet** [comment]
Right, just realised that I can set the range from a specific device. Thank you. 
I have another issue which I hope you can help me with. I am running a system with two GPUs. Apparently by running Device.firstGPU() and Device.best() I can get both of my GPUs (it considers the second to be the best). 

Now, I have a problem of size N. I've tried creating two different ranges and kernels in order to split the program between the two GPUs. One range and kernel for GPU1 and one for GPU2. 

The issue I'm having is that when I try to execute those two kernels like this: 
kernel1.execute(range1);
kernel2.execute(range2);
each GPU seems to be getting utilized by only 50%. However, if I run the whole problem size on a single Kernel, the one GPU that is used is utilized by 100%.

Would you have any idea what could be causing this issue and how I could resolve it?

Thanks again.

**Comment by grfrost** [comment]
Do the kernels share the same buffer/array?  Or did you partition the data between two sets of input and output data?

Also note that 
kernel1.execute(range1) 
kernel2.execute(range2)

Will run them sequentially. 

Consider 
(new Thread(new Runnable(){void run(){kernel1.execute(range1);}})).start()
(new Thread(new Runnable(){void run(){kernel2.execute(range2);}})).start()

To run them in parallel.

Gary

**Comment by harrydet** [comment]
They share the same buffer. Does this mean that they get exclusive lock on the array they are using? If so I suppose I could fix my issue by splitting the inputs to two arrays?

Also I don't mind them being run sequentially at the moment but thanks for the suggestion.

**Comment by grfrost** [comment]
Actually if they share the same buffer you are unnecessarily copying the whole buffer in/out and 'you will have to' run them sequentially ;) 

Unless you have been 'cunning' whilst declaring your ranges... ;)

**Comment by harrydet** [comment]
I've basically created the ranges using 

Range range1 = gpu1.createRange(N/2);
Range range2 = gpu2.createRange(N/2);

And inside kernel2 I have this line of code to determine which item is ran

int gid = getGlobalId() + N/2;

So I guess they are both running what they should be running.

**Comment by harrydet** [comment]
Tried running it with 2 threads, it worked :) 
Thanks a lot, closing.

**Comment by niinja2** [comment]
Hi i have the same problem of picking GPU-s
This code:

Device chosen=null;
for (Device device: devices.getAll()){
if (device.getVendor().contains("AMD") && device.isGPU()){
chosen = device;
break;
}
} 

does not work for me. It cannot recognize "devices". Is it supposed to be a list of all devices? how do i create it? My first GPU and best GPU are the same so i have a problem because there is no way of picking second or other GPU-s.

**Comment by harrydet** [comment]
Yea that's the same problem I was having. If your first and best GPU are the same maybe try swapping them around in the motherboard? Not sure if that's gonna work though...

**Comment by niinja2** [comment]
Tried it , it did not work, but i used

    Device rad7790= Device.best();
        OpenCLDevice rad5670 = OpenCLDevice.select(new
OpenCLDevice.DeviceComparitor(){
               public OpenCLDevice select(OpenCLDevice lhs, OpenCLDevice
rhs){

return(lhs.getMaxMemAllocSize()<rhs.getMaxMemAllocSize()?lhs:rhs);

               }
            });
        System.out.println(rad7790+"\n "+rad5670);

This selects GPU based on size of global memory.
